Add an on_pickup callback to items

DS-Minetest opened the pull request:
The changes in `lua_api.txt` should say enough.

DS-Minetest commented on doc/lua_api.txt:
````diff
@@ -4239,6 +4239,28 @@ Definition tables
+			self.object:remove()
+		end
+		```
+		]]
````
Huh, shall in `lua_api.txt` be indented with spaces?

nerzhul replied:
yes

paramat commented:
Maybe #3962 #4035 #5647 could be useful to look at?

DS-Minetest commented:
@paramat I don't understand. This PR has nothing directly to do with inventories.

nerzhul commented on builtin/game/item_entity.lua:
```diff
@@ -207,7 +207,12 @@ core.register_entity(":__builtin:item", {
 
-	on_punch = function(self, hitter)
+	on_punch = function(self, hitter, ...)
+		local item_def = core.registered_items[self.itemstring]
```
if you want to do this in Lua, all this added block should be in the next condition (if inv and self.itemstring != "")

paramat commented on doc/lua_api.txt:
```diff
@@ -4239,6 +4239,28 @@ Definition tables
+        on_pickup = func(item_luaentity, up_picker, time_from_last_punch, tool_capabilities, dir),
+        --[[
+        ^ Called when a dropped item is punched and shall be picked up.
+        ^ Shall remove dropped item and add itemstack to `up_picker`'s inventory.
```
Why document this if it is already happening, this callback doesn't make this happen.
`up_picker` is bad English, how about `picker` ?

paramat replied:
How about
`Called when a dropped item is punched and shall be picked up by the 'picker'`
?

paramat commented on doc/lua_api.txt:
```diff
@@ -4239,6 +4239,28 @@ Definition tables
         ^ default: minetest.item_drop ]]
+        on_pickup = func(item_luaentity, up_picker, time_from_last_punch, tool_capabilities, dir),
+        --[[
+        ^ Called when a dropped item is punched and shall be picked up.
```
'and shall be' is not good English, do you mean to explain this is called before the item is added to the inventory?

paramat commented:
> This PR has nothing directly to do with inventories.

Sorry, my mistake.

DS-Minetest commented:
I think, it's good now.

paramat commented on doc/lua_api.txt:
```diff
@@ -2447,6 +2447,11 @@ Call these functions only at load time!
+* `minetest.register_on_item_pickup(func(item_luaentity, picker, time_from_last_punch, tool_capabilities, dir))`
+    * Called when item is picked up by `picker`.
+    * Newest functions are called first.
+    * If `true` is returned, the following functions are not called.
```
What 'following functions'?

paramat commented on doc/lua_api.txt:
```diff
@@ -2447,6 +2447,11 @@ Call these functions only at load time!
+    * Called when item is picked up by `picker`.
+    * Newest functions are called first.
+    * If `true` is returned, the following functions are not called.
+    * If `false` is returned, the following functions are not called and the item is not picked up.
```
Same here.

paramat commented on doc/lua_api.txt:
```diff
@@ -4239,6 +4244,32 @@ Definition tables
+        on_pickup = func(item_luaentity, picker, time_from_last_punch, tool_capabilities, dir),
+        --[[
+        ^ Called when a dropped item is punched and picked up by `picker`.
+        ^ Shall remove dropped item and add itemstack to `picker`'s inventory or
```
Do you mean that the function should either return 'true', or itself do the job of removing the object and adding it to the picker's inventory (instead of existing code)?

'Shall' is unclear English, 'should' is better, and a comma is good for scope:
`Should remove dropped item and add itemstack to `picker`'s inventory, or return `true`.`

paramat commented on doc/lua_api.txt:
```diff
@@ -4239,6 +4244,32 @@ Definition tables
+        ^ `item_luaentity` is the luaentity of the dropped item.
+        ^ The `picker` may be any `ObjectRef` or `nil`.
+        ^ Not setting it or returning `true` will first call all functions that
+          are register on pickup (see `minetest.register_on_item_pickup`) and
```
registered

paramat commented on doc/lua_api.txt:
```diff
@@ -4239,6 +4244,32 @@ Definition tables
+          return `true`.
+        ^ `item_luaentity` is the luaentity of the dropped item.
+        ^ The `picker` may be any `ObjectRef` or `nil`.
+        ^ Not setting it or returning `true` will first call all functions that
```
Not setting what, needs to be clearer.

paramat commented:
Documentation is unclear and awkward English, which may not be your first language so that's understandable.

DS-Minetest commented:
I'll move `register_on_item_pickup` to `register.lua`. >_<

paramat commented:
Thanks, docs are clearer now.

rubenwardy commented on builtin/game/item_entity.lua:
```diff
@@ -207,9 +207,29 @@ core.register_entity(":__builtin:item", {
+			-- Call every function registered on pickup.
+			local success
+			for i = 1, #core.registered_on_item_pickups do
+				success = core.registered_on_item_pickups[i](self, hitter, ...)
```
success is not used again, so should put local here

rubenwardy commented on builtin/game/item_entity.lua:
```diff
@@ -207,9 +207,29 @@ core.register_entity(":__builtin:item", {
+					not item_def.on_pickup(self, hitter, ...) then
+				return
+			end
+			-- Call every function registered on pickup.
```
please put an empty new line above here

rubenwardy commented on builtin/game/item_entity.lua:
```diff
@@ -207,9 +207,29 @@ core.register_entity(":__builtin:item", {
+					return
+				end
+			end
+			-- Do the default case.
```
please put an empty new line above here

rubenwardy commented on builtin/game/item_entity.lua:
```diff
@@ -207,9 +207,29 @@ core.register_entity(":__builtin:item", {
+			local inv = hitter:get_inventory()
+			if not inv then
+				return
+			end
```
please put an empty new line below here

rubenwardy commented on doc/lua_api.txt:
```diff
@@ -2447,6 +2447,12 @@ Call these functions only at load time!
 * `minetest.register_on_item_eat(func(hp_change, replace_with_item, itemstack, user, pointed_thing))`
     * Called when an item is eaten, by `minetest.item_eat`
     * Return `true` or `itemstack` to cancel the default item eat response (i.e.: hp increase)
+* `minetest.register_on_item_pickup(func(item_luaentity, picker, time_from_last_punch, tool_capabilities, dir))`
```
`item_luaentity` -> `luaentity`

I'm also wondering if giving the ObjectRef is a better API. Hmmm

rubenwardy commented on doc/lua_api.txt:
```diff
@@ -4239,6 +4245,32 @@ Definition tables
+        on_pickup = func(item_luaentity, picker, time_from_last_punch, tool_capabilities, dir),
+        --[[
+        ^ Called when a dropped item is punched and picked up by `picker`.
+        ^ Function should remove dropped item and add itemstack to `picker`'s
```
should maybe be: Return false to cancel the pick up.

rubenwardy commented on doc/lua_api.txt:
```diff
@@ -4239,6 +4245,32 @@ Definition tables
+          inventory, or return `true`.
+        ^ `item_luaentity` is the luaentity of the dropped item.
+        ^ The `picker` may be any `ObjectRef` or `nil`.
+        ^ Not setting this function or returning `true` will first call all functions
```
This function is called before running register_on_item_pickup callbacks, and returning false will cancel their execution

rubenwardy commented on doc/lua_api.txt:
````diff
@@ -4239,6 +4245,32 @@ Definition tables
+          that are registered on pickup (see `minetest.register_on_item_pickup`)
+          and then normally result in:
+          ```lua
+          function(self, hitter)
````
this sort of snippet isn't usually given

rubenwardy commented:
willing to work on this?

paramat commented:
@DS-Minetest , needs attention, will close in 12 days (1 month after reviewer requests) if no progress.

paramat commented:
Neglected, closing.

DS-Minetest commented:
Sorry for my absence. I rebased, fixed rubenwardy's small things and pushed to the PR branch, but it doesn't show here. Can this be reopened or should I make a new PR?
